docs/unread_messages.md

```diff
@@ -0,0 +1,32 @@
+# Unread message synchronization
+
+In general displaying unread counts for all streams and topics may require
+downloading an unbounded number of messages. Consider a user who has a muted
+stream or topic and has not read the backlog in a month; to have an accurate
+unread count we would need to load all messages this user has received in the
+past month. This is inefficient for web clients and even more for mobile
+devices.
+
+We work around this by including a list of unread message ids in the initial
+state grouped by recipient and subject. This data is included in the
+`unread_msgs` key if both update_message_flags and message are required in the
+register call.
+
+```
+[
+    [
+        {
+            'display_recipient': 'general',
+            'recipient_id': 54,
+            'subject': 'lunch',
+        },
+        [400, 823]
```

Comment:
shape-wise, maybe we want to move the message IDs into a `message_ids` list inside the dict (and then get rid of the inner list), just to have a somewhat cleaner and more extensible interface?

We'll also eventually want to add a `max_message_id` field for recent threads (even those without unread messages), since that'll be an important piece of a model, which would also be easy to add in the dict model.

Does that seem reasonable @armooo?